### PR TITLE
Storage rework

### PR DESCRIPTION
**Pull request by bdlukaa**
Part of https://github.com/supabase-community/supabase-dart/issues/109

- Now errors should be handled inside of a `try/catch` block
- Added `StorageFileApi.copy` and `StorageFileApi.createSignedUrls`
- Renamed `fetch` to `storageFetch`, in order to avoid conflicts between other supabase libraries
- Improved documentation on methods that already had docs
- Added `lints` for linting

### CODE REVIEW

File: lib/src/storage_file_api.dart [hunk 1]
```diff
@@ -4,302 +4,275 @@ import 'package:storage_client/src/fetch.dart';
...
+      '$url/object/copy',
+      {
+        'bucketId': bucketId,
+        'sourceKey': fromPath,
+        'destinationKey': toPath,
+      },
+      options: options,
+    );
+    return (response as Map<String, dynamic>)['message'] as String;
+  }
```
Comment (bdlukaa) [inline lib/src/storage_file_api.dart:157]: Reference: https://github.com/supabase/storage-js/blob/main/src/lib/StorageFileApi.ts#L186-L207

File: lib/src/storage_file_api.dart [hunk 2]
```diff
@@ -4,302 +4,275 @@ import 'package:storage_client/src/fetch.dart';
...
+      options: options,
+    );
+    final urls = (response as List).map((e) {
+      return '$url${e['signedUrl']}';
+      // return {
+      //   ...e,
+      //   'signedUrl': '$url${e['signedUrl']}',
+      // };
+    });
+    return urls.toList();
```
Comment (bdlukaa) [inline lib/src/storage_file_api.dart:208]: I don't know if this is correct, @phamhieu .

https://github.com/supabase/storage-js/blob/main/src/lib/StorageFileApi.ts#L239-L269

File: lib/src/types.dart [hunk 3]
```diff
@@ -104,49 +101,68 @@ class SearchOptions {
...
+  const StorageException(this.message, {this.error, this.statusCode});
 
-  StorageError.fromJson(dynamic json)
-      : assert(json is Map<String, dynamic>),
-        message = (json as Map<String, dynamic>)['message'] as String,
+  StorageException.fromJson(Map<String, dynamic> json, [String? statusCode])
+      : message = json['message'] as String,
         error = json['error'] as String?,
-        statusCode = json['statusCode'] as String?;
+        statusCode = (json['statusCode'] as String?) ?? statusCode;
```
Comment (bdlukaa) [inline lib/src/types.dart:162]: the fallback status code helps debugging

File: lib/src/fetch.dart [hunk 4]
```diff
@@ -7,7 +7,7 @@ import 'package:mime/mime.dart';
 import 'package:storage_client/src/types.dart';
 import 'package:universal_io/io.dart';
 
-Fetch fetch = Fetch();
+Fetch storageFetch = Fetch();
```
Comment (bdlukaa) [inline lib/src/fetch.dart:10]: This was renamed to `storageFetch` in order to not have conflicts with other supabase libraries, such as postgrest.

**Review by dshukertjr — COMMENTED** [review]
I love this rework. 

I especially love using try/catch to handle errors, but I was wondering how this would translate over on postgrest-dart. Would we be simply getting rid of `error` property on [`PostgrestResponse`](https://github.com/supabase-community/postgrest-dart/blob/master/lib/src/postgrest_response.dart#L4) object? Did you have any better idea?

**Comment by bdlukaa** [comment]
@dshukertjr Yes! I'm looking for a solution to remove postgrest response

File: lib/src/storage_file_api.dart (hunk 1 above)
Comment (dshukertjr) [reply lib/src/storage_file_api.dart:157]: I love the implementation!

File: lib/src/storage_file_api.dart [hunk 5]
```diff
@@ -4,302 +4,275 @@ import 'package:storage_client/src/fetch.dart';
...
+  /// [expiresIn] is the number of seconds until the signed URLs expire. For
+  /// example, `60` for URLs which are valid for one minute.
+  Future<List<String>> createSignedUrls(
+    List<String> paths,
+    int expiresIn,
+  ) async {
+    final options = FetchOptions(headers: headers);
+    final response = await storageFetch.post(
+      '$url/object/sign/$bucketId',
+      {'expiresIn': expiresIn},
```
Comment (dshukertjr) [inline lib/src/storage_file_api.dart:198]: Do we need to pass the `paths` as the body as well here?
```suggestion
      {
        'expiresIn': expiresIn,
        'paths': paths,
      },
```

https://github.com/supabase/storage-js/blob/main/src/lib/StorageFileApi.ts#L256

**Review by dshukertjr — COMMENTED** [review]
>Yes! I'm looking for a solution to remove postgrest response

@bdlukaa Amazing! looking forward to it. I think I found one little thing within `createSignedUrls`, but other than that, it LGTM!

**Review by dshukertjr — APPROVED** [review]
LGTM! Such an amazing work!

**Pull request merged**